barretenberg/cpp/src/barretenberg/commitment_schemes/commitment_key.hpp

```diff
@@ -79,20 +81,30 @@ template <class Curve> class CommitmentKey {
      * @param polynomial a univariate polynomial p(X) = ∑ᵢ aᵢ⋅Xⁱ
      * @return Commitment computed as C = [p(x)] = ∑ᵢ aᵢ⋅Gᵢ
      */
-    Commitment commit(std::span<const Fr> polynomial)
+    Commitment commit(PolynomialSpan<const Fr> polynomial)
     {
         BB_OP_COUNT_TIME();
-        // See constructor, we must round up the number of used srs points to a power of 2.
-        const size_t consumed_srs = numeric::round_up_power_2(polynomial.size());
+        // WORKTODO(sparse) rework this - should we always inform how much SRS we need on-demand?
+        // We must have a power-of-2 SRS points *after* subtracting by start_index.
+        const size_t consumed_srs = numeric::round_up_power_2(polynomial.size()) + polynomial.start_index;
+        auto srs = crs_factory->get_prover_crs(consumed_srs);
+        // We only need the
         if (consumed_srs > srs->get_monomial_size()) {
-            info("Attempting to commit to a polynomial that needs ",
-                 consumed_srs,
-                 " points with an SRS of size ",
-                 srs->get_monomial_size());
-            ASSERT(false);
+            throw_or_abort(format("Attempting to commit to a polynomial that needs ",
```

Comment:
I don't see the utility of continuing on. Actually, there is security risk whenever we knowingly read past allocated memory.